Handled referrer attribution capture and storage

rshbhgrg opened the pull request:
refs https://github.com/TryGhost/Team/issues/1906, https://github.com/TryGhost/Team/issues/1907

- updates member attribution script to capture referrer url/source/medium information and store it in the history for later usage
  - stores `refUrl`, `refSource` and `refMedium` in history
- updates attribution service to use referrer data from history, and translate it to referrer source and medium using `ReferrerTranslator`
- includes referrer source/medium/url to stripe checkout metadata, allowing storing it in event later when a member completes payment

SimonBackx commented on ghost/member-attribution/lib/attribution.js:
```diff
@@ -80,21 +88,52 @@ class AttributionBuilder {
+
+        for (const item of history) {
+            const {referrerSource, referrerMedium, referrerUrl} = this.referrerTranslator.getAttributionData(item);
+            if (referrerSource && referrerMedium) {
+                return {
+                    referrerSource,
+                    referrerMedium,
+                    referrerUrl
+                };
+            }
```
I'm wondering here that when both `item.referrerSource` and `item.referrerMedium` are null/undefined, that we still get default values. So if you visit
`https://ghost.site/my-post?rel=this-is-referred`
Then later click through to a different post
`https://ghost.site/other-post`

You'll essentially lose that first `rel` because the last one already will return some source and medium defaults. Is this intended behaviour?

rishabhgrg replied:
@SimonBackx Yea i see what you mean, we'll definitely update this to be more smart about how we calculate the referrer source/medium from the history, for example as you said we could maybe first check history for any entry that has `ref` and if not only then fallback to calculating them from URL. 
The current logic to calculate referrer sources just places the pieces in place for where things will happen, and we'll have more iterations on how the values are calculated from history once this is good :)

SimonBackx replied:
Perfect! 👌 Just wanted to make sure 😊

rishabhgrg replied:
That's good, thanks 😄

rishabhgrg replied:
This should be much better now with updated logic to calculate referrer from history :)

SimonBackx commented on ghost/core/core/frontend/src/member-attribution/member-attribution.js:
```diff
@@ -71,6 +71,26 @@ const LIMIT = 15;
+        try {
+            // Fetch source/medium from query param
+            const url = new URL(window.location.href);
+            refParam = url.searchParams.get('ref');
+            sourceParam = url.searchParams.get('source');
+            utmSourceParam = url.searchParams.get('utm_source');
+            utmMediumParam = url.searchParams.get('utm_medium');
```
Should we remove these params from the URL afterwards?

rishabhgrg replied:
🤔 I was thinking about it too, specially after looking at the fact we remove it for attribution id/sources, but i think these params are different because they are set externally, and there might be other analytics/tools that are setup to rely on them, so imo doesn't make sense for us to remove them, wdyt ?

SimonBackx replied:
You are right about external tools, didn't think about that 🤔

rishabhgrg replied:
Will keep it in for now, and based on how it feels later on testing we can remove it if needed

allouis commented on ghost/member-attribution/lib/referrer-translator.js:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    getAttributionData({referrerUrl, referrerSource, referrerMedium}) {
+        return {
```
Does it make more sense to return `null` here when there is no source or medium?

That way this class is solely responsible for generating referrer data, including what is/isn't a valid referrer

rishabhgrg replied:
Yeah that makes sense, though I am wondering if this should change completely where instead of calculating referrer for each individual history item, we might as well pass in the whole history here so it can have some smart calculations on how to derive the `Source` and `Medium` based on different history items. Gonna try that, otherwise switch to returning `null` if it keeps the current check for individual history item

rishabhgrg replied:
@allouis Updated the logic here to pass in the `history` directly, as the logic to smartly calculate referrer information needs access to whole history so it can be looped over for different cases. Also now returns `null` instead of object with nulls when no referrer information can be created, lmk if this makes sense.

allouis reviewed (commented):
Approach looks good - just one main comment about moving the logic of what is/isn't a valid referrer into a single place

rishabhgrg commented on ghost/member-attribution/lib/referrer-translator.js:
```diff
@@ -0,0 +1,95 @@
+        this.urlUtils = urlUtils;
+    }
+
+    getAttributionData({referrerUrl, referrerSource, referrerMedium}) {
```
Yeah right, will update JSDoc types, wanted to make sure pieces are in right place before going all in with those changes :)

rishabhgrg commented on ghost/member-attribution/lib/referrer-translator.js:
```diff
@@ -0,0 +1,95 @@
+     * @param {Object} deps.urlUtils
+     */
+    constructor({urlUtils}) {
+        this.urlUtils = urlUtils;
```
Yea that works as well, the reason I had left `urlUtils` was in case we'd want to extend the logic in future, will switch to simpler `adminUrl` for now and we can switch back again later. 👍

rishabhgrg replied:
Updated deps to pass in `adminUrl` and `siteUrl` directly ✔️

rishabhgrg commented on ghost/member-attribution/lib/attribution.js:
```diff
@@ -15,11 +15,19 @@ class Attribution {
      * @param {'page'|'post'|'author'|'tag'|'url'} [data.type]
+     * @param {string|null} [data.referrerSource]
+     * @param {string|null} [data.referrerMedium]
+     * @param {string|null} [data.referrerUrl]
```
Done

rishabhgrg commented on ghost/member-attribution/lib/referrer-translator.js:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    isGhostOrgUrl(url) {
+        if (url.includes('ghost.org')) {
```
Updated logic to use URL checks where possible instead of string, thanks for the idea.

rishabhgrg commented:
@allouis @SimonBackx Thanks for the feedbacks. Have updated PR now with suggested changes, including logic cleanup where `ReferrerTranslator` now takes in the `history` directly instead of each history item so it can calculate the referrer information more effectively. This is missing a piece where we use a curated list of external urls for calculating source/medium in the translator, but thats incoming shortly after as a follow-up PR.

rshbhgrg committed:
Updated attribution script to capture referrer information

refs TryGhost/Team#1906

- captures referrer url and any source/medium param for referrer
- maintains history with latest referrer information when moving around pages

rshbhgrg committed:
Updated attribution service to handle referrer information

refs TryGhost/Team#1907

- calculates final attribution source and medium using captured referrer information in history
- adds new referrer-translator that goes through available history and based to determine most valid referrer info
- includes referrer url, source and medium in the attribution data for storage

rshbhgrg committed:
Updated stripe checkout session metadata for referrer data

- adds referrer source, medium and url to stripe checkout metadata for later storage
